File: golang/positions_test.go
package flyql

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"regexp"
	"strconv"
	"strings"
	"testing"
)

var fixtureIndexRE = regexp.MustCompile(`^([^\[\]]+)(?:\[(\d+)\])?$`)

// getByPath traverses a Node by dot-notation path, handling [i] index suffixes.
// Field-name translation mirrors the shared JSON fixture convention:
//   - snake_case paths are converted to CamelCase (e.g. "bool_operator_range" → "BoolOperatorRange")
func getByPath(root *Node, path string) (any, error) {
	parts := strings.Split(path, ".")
	if len(parts) > 0 && parts[0] == "root" {
		parts = parts[1:]
	}
	var current any = root
	for _, part := range parts {
		m := fixtureIndexRE.FindStringSubmatch(part)
		if m == nil {
			return nil, fmt.Errorf("invalid path segment: %q", part)
		}
		name := snakeToCamel(m[1])
		idxStr := m[2]
		v := reflect.ValueOf(current)
		for v.Kind() == reflect.Ptr {
			if v.IsNil() {
				return nil, fmt.Errorf("nil pointer at segment %q", part)
			}
			v = v.Elem()
		}
		if v.Kind() != reflect.Struct {
			return nil, fmt.Errorf("cannot take field %q from %v (kind=%s)", name, current, v.Kind())
		}
		f := v.FieldByName(name)
		if !f.IsValid() {
			return nil, fmt.Errorf("field %q not found (path=%q)", name, part)
		}
		current = f.Interface()
		if idxStr != "" {
			idx, _ := strconv.Atoi(idxStr)
			cv := reflect.ValueOf(current)
			if cv.Kind() != reflect.Slice && cv.Kind() != reflect.Array {
				return nil, fmt.Errorf("not a slice at %q", part)
			}
			current = cv.Index(idx).Interface()
		}
	}
	return current, nil
}

func snakeToCamel(s string) string {
	parts := strings.Split(s, "_")
	out := ""
	for _, p := range parts {
		if p == "" {
			continue
		}
		out += strings.ToUpper(p[:1]) + p[1:]
	}
	return out
}

type fixtureCase struct {
	Name           string                 `json:"name"`
	Input          string                 `json:"input"`
	ExpectedResult string                 `json:"expected_result"`
	ExpectedRanges map[string][]int       `json:"expected_ranges"`
	ExpectedError  map[string]interface{} `json:"expected_error"`
}

type fixtureFile struct {
	TestSuite string        `json:"test_suite"`
	Tests     []fixtureCase `json:"tests"`
}

func findFixturesDir(t *testing.T) string {
	// Walk up from cwd looking for tests-data/core/parser/positions
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}
	for i := 0; i < 5; i++ {
		candidate := filepath.Join(wd, "tests-data", "core", "parser", "positions")
		if _, err := os.Stat(candidate); err == nil {
			return candidate
		}
		wd = filepath.Dir(wd)
	}
	t.Fatalf("could not locate tests-data/core/parser/positions")
	return ""
}

func TestPositions(t *testing.T) {
	dir := findFixturesDir(t)
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("readdir: %v", err)
	}
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		data, err := os.ReadFile(filepath.Join(dir, e.Name()))
		if err != nil {
			t.Fatalf("read %s: %v", e.Name(), err)
		}
		var ff fixtureFile
		if err := json.Unmarshal(data, &ff); err != nil {
			t.Fatalf("unmarshal %s: %v", e.Name(), err)
		}
		for _, tc := range ff.Tests {
			tc := tc
			t.Run(fmt.Sprintf("%s::%s", e.Name(), tc.Name), func(t *testing.T) {
				p := NewParser()
				err := p.Parse(tc.Input)
				if tc.ExpectedResult == "error" {
					if err == nil {
						t.Fatalf("expected error, got none")
					}
					pe, ok := err.(*ParseError)
					if !ok {
						t.Fatalf("expected *ParseError, got %T: %v", err, err)
					}
					wantErrno := int(tc.ExpectedError["errno"].(float64))
					if pe.Code != wantErrno {
						t.Fatalf("errno: got %d, expected %d", pe.Code, wantErrno)
					}
					rangeArr := tc.ExpectedError["range"].([]interface{})
					want := Range{Start: int(rangeArr[0].(float64)), End: int(rangeArr[1].(float64))}
					if pe.Range != want {
						t.Fatalf("error range: got %+v, expected %+v", pe.Range, want)
					}
					return
				}
				if err != nil {
					t.Fatalf("parse error: %v", err)
				}
				for path, expected := range tc.ExpectedRanges {
					got, perr := getByPath(p.Root, path)
					if perr != nil {
						t.Fatalf("%s: %v", path, perr)
					}
					want := Range{Start: expected[0], End: expected[1]}
					// unwrap *Range
					if gptr, ok := got.(*Range); ok {
						if gptr == nil {
							t.Fatalf("%s: got nil *Range, expected %+v", path, want)
						}
						got = *gptr
					}
					if !reflect.DeepEqual(got, want) {
						t.Fatalf("%s: got %+v, expected %+v", path, got, want)
					}
				}
			})
		}
	}
}

File: javascript/packages/flyql/test/core/positions.test.js
import { describe, it, expect } from 'vitest'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { Parser, ParserError } from '../../src/index.js'
import { Range } from '../../src/core/range.js'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const FIXTURES_DIR = path.join(__dirname, '..', '..', '..', '..', '..', 'tests-data', 'core', 'parser', 'positions')

const INDEX_RE = /^([^[\]]+)(?:\[(\d+)\])?$/

function snakeToCamel(s) {
    return s.replace(/_([a-z])/g, (_, c) => c.toUpperCase())
}

function getByPath(root, pathStr) {
    let current = root
    const parts = pathStr.split('.')
    const start = parts[0] === 'root' ? 1 : 0
    for (let i = start; i < parts.length; i++) {
        const m = parts[i].match(INDEX_RE)
        if (!m) throw new Error(`invalid path segment: ${parts[i]}`)
        const name = snakeToCamel(m[1])
        const idx = m[2]
        if (current === null || current === undefined) {
            throw new Error(`null/undefined at segment ${parts[i]}`)
        }
        current = current[name]
        if (idx !== undefined) {
            current = current[parseInt(idx, 10)]
        }
    }
    return current
}

function loadFixtures() {
    const files = fs
        .readdirSync(FIXTURES_DIR)
        .filter((f) => f.endsWith('.json'))
        .sort()
    const cases = []
    for (const file of files) {
        const data = JSON.parse(fs.readFileSync(path.join(FIXTURES_DIR, file), 'utf-8'))
        for (const tc of data.tests) {
            cases.push({ file, tc })
        }
    }
    return cases
}

describe('AST source positions', () => {
    for (const { file, tc } of loadFixtures()) {
        it(`${file}::${tc.name}`, () => {
            if (tc.expected_result === 'error') {
                let caught = null
                try {
                    const p = new Parser()
                    p.parse(tc.input)
                } catch (e) {
                    caught = e
                }
                expect(caught).toBeInstanceOf(ParserError)
                expect(caught.errno).toBe(tc.expected_error.errno)
                const want = tc.expected_error.range
                expect(caught.range).toEqual(new Range(want[0], want[1]))
                return
            }
            const p = new Parser()
            p.parse(tc.input)
            for (const [pathStr, expected] of Object.entries(tc.expected_ranges)) {
                const actual = getByPath(p.root, pathStr)
                expect(actual, `path ${pathStr}`).not.toBeNull()
                expect(actual, `path ${pathStr}`).toEqual(new Range(expected[0], expected[1]))
            }
        })
    }
})
